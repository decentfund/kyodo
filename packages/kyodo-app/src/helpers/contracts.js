import MembersContract from '@kyodo/contracts/build/contracts/MembersV2.json';
import PeriodsContract from '@kyodo/contracts/build/contracts/PeriodsV1.json';
import DomainsContract from '@kyodo/contracts/build/contracts/DomainsV1.json';

const generateContractConfig = ({
  contractName,
  contractAbi,
  web3,
  address,
}) => {
  const contractConfig = {
    contractName,
    web3Contract: new web3.eth.Contract(contractAbi, address),
  };
  return contractConfig;
};

const generateContractConfigFromEvent = ({ event, web3, address }) => {
  let contractName, contractAbi;
  if (event === 'PeriodsAddressChanged') {
    contractName = 'Periods';
    contractAbi = PeriodsContract.abi;
  } else if (event === 'MembersAddressChanged') {
    contractName = 'Members';
    contractAbi = MembersContract.abi;
  } else if (event === 'DomainsAddressChanged') {
    contractName = 'Domains';
    contractAbi = DomainsContract.abi;
  }

  if (contractName) {
    return generateContractConfig({ contractName, contractAbi, web3, address });
  }
  return;
};

const generateContractConfigFromName = ({
  name: contractName,
  web3,
  address,
}) => {
  let contractAbi;
  if (contractName === 'Periods') {
    contractAbi = PeriodsContract.abi;
  } else if (contractName === 'Members') {
    contractAbi = MembersContract.abi;
  }

  if (contractAbi) {
    return generateContractConfig({ contractName, contractAbi, web3, address });
  }
  return;
};

export default generateContractConfig;
export { generateContractConfigFromEvent, generateContractConfigFromName };
