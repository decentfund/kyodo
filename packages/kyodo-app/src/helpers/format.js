const ETH = 'ETH';
const DAI = 'DAI';
const USD = 'USD';
const EUR = 'EUR';

export function formatCurrency(value, currency, precision = 0) {
  let decimalPoints = precision;
  if (!precision) {
    switch (currency) {
      case USD:
      case EUR:
      case DAI:
        decimalPoints = 2;
        break;
      case ETH:
        decimalPoints = 4;
        break;
      default:
        decimalPoints = 0;
    }
  }
  return parseFloat(
    Math.round(value * 10 ** decimalPoints) / 10 ** decimalPoints,
  );
}

export const formatEth = value => `${formatCurrency(value, ETH)} ETH`;
export const formatEur = value => `${formatCurrency(value, EUR)} â‚¬`;

export const formatDecimals = (value, decimals) =>
  value / Math.pow(10, decimals);
