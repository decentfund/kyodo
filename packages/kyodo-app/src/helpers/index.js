import web3 from 'web3';

export const isValidAddress = address => web3.utils.isAddress(address);
