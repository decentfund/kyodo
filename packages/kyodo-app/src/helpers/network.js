export const getNetworkName = (network) => {
  switch (network) {
    case 1: return 'Mainnet';
    case 2: return 'Morden';
    case 3: return 'Ropsten';
    case 4: return 'Rinkeby';
    case 42: return 'Kovan';
    case 5777: return 'Private';
    default: return 'Development';
  }
};
