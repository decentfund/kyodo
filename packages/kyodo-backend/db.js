const mongoose = require('mongoose');
const { SystemError } = require('./errors');

const Schema = mongoose.Schema;

const Period = require('./models/Period');
const User = require('./models/User');
const Tip = require('./models/Tip');

mongoose.connection.on('error', e => {
  if (e.message.code === 'ETIMEDOUT') {
    console.log(e);
  }
  console.log(e);
});

mongoose.connection.once('open', () => {
  console.log(`MongoDB successfully connected to mongoUri`);
});

const initDb = () => mongoose.connect('mongodb://localhost/colony');

const taskSchema = new mongoose.Schema({
  taskId: String,
  taskTitle: String,
  specificationHash: String,
  finalized: Boolean,
  cancelled: Boolean,
  dateCreated: Date,
  dueDate: Date,
  potId: Number,
  // domainId: Number,
  // id: Number,
  // skillId: Number
});

const colonySchema = new mongoose.Schema({
  colonyName: String,
  colonyId: String,
  colonyAddress: String,
  tokenAddress: String,
  tokenName: String,
  tokenSymbol: String,
  creationBlockNumber: Number,
  creationDate: Date,
  periodIds: Array,
});

colonySchema.virtual('currentPeriodId').get(function() {
  return this.periodIds[this.periodIds.length - 1];
});

const domainSchema = new Schema({
  domainId: Number,
  domainTitle: String,
  localSkillId: Number,
  potId: Number,
});

const Task = mongoose.model('Task', taskSchema);
const Colony = mongoose.model('Colony', colonySchema);
const Domain = mongoose.model('Domain', domainSchema);

const getColonyById = async colonyId => {
  return await Colony.findOne({ colonyId });
};

const getCurrentUserPeriod = async (alias, periodId) => {
  const user = await User.findOne({ alias });
  const period = await Period.findOne({ user: user._id, periodId });
  const sentTips = await Tip.find({ from: user, periodId });
  const usedPoints = sentTips.reduce((sum, v) => sum + v.amount, 0);

  if (!period) throw new SystemError('Cannot find user current period');

  return {
    ...period,
    initialBalance: period.balance,
    balance: period.balance - usedPoints,
  };
};

const getDomainByCode = async code => {
  const domain = await Domain.findOne({ domainTitle: code });
  if (!domain) throw Error('Domain not found');

  return domain;
};

module.exports = {
  initDb,
  Task,
  Colony,
  Domain,
  Tip,
  User,
  Period,
  getColonyById,
  getCurrentUserPeriod,
  getDomainByCode,
};
