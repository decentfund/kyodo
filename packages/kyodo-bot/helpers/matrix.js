const isDirectChat = (room, user) => {
  let type = "room";
  // if (type === 'directMessage') {
  // state.sources[roomId].name = room.name.replace(/@(\w+):.+/, '$1');
  // }
  const allMembers = room.currentState.getMembers();
  if (
    type === "room" &&
    allMembers.length <= 2 &&
    allMembers.some(m => m.userId === user)
  ) {
    // if (
    // // allMembers.some(m => m.getDMInviter()) &&
    // ) {
    type = "directMessage";
    // }
  }

  return type === "directMessage";
};

export const getOrCreatePrivateRoom = async (client, event, alias = null) => {
  console.log("getting");
  const party = alias || event.getSender();

  let userRoom = client.getRooms().find(r => isDirectChat(r, party));
  let roomId;
  if (userRoom) roomId = userRoom.roomId;
  if (!userRoom) {
    const newChat = await client.createRoom({
      preset: "trusted_private_chat",
      invite: [party],
      is_direct: true,
    });
    roomId = newChat.room_id;
  }

  return roomId;
};
