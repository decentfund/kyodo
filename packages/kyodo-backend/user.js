import web3 from 'web3';
import { Colony, getColonyById, getCurrentUserPeriod } from './db';
import User from './models/User';
import { createAndSaveNewUserPeriod } from './period';
const Point = require('./models/Point');
const { ModelError } = require('./errors');

export const dbAddUser = async ({
  address,
  alias,
  balance,
  domains,
  tasks,
}) => {
  const user = new User({
    address,
    alias,
    balance,
    domains,
    tasks,
    dateCreated: Date.now(),
  });

  await user.save(err => {
    if (err) return null;
  });

  return user;
};

export const addUser = async ({ alias }) => {
  const colony = await Colony.findOne();
  const user = await dbAddUser({ alias });
  await createAndSaveNewUserPeriod({
    periodId: colony.currentPeriodId,
    balance: 0,
    user,
  });
  return user;
};

export const dbGetAllUsers = async () => {
  return await User.find((err, users) => {
    if (err) {
      console.log(err);
      return [];
    }
    return users;
  });
};

export const getAllUsers = async (req, res) => {
  let users = await dbGetAllUsers();
  res.status(200).send(users);
  return users;
};

export const getUserByAddress = async (req, res) => {
  await User.find({ address: req.body.address });
};

export const getUserByAlias = async (req, res) => {
  await User.find({ alias: req.body.alias });
};

export const dbGetUserByAlias = async alias => {
  const user = await User.findOne({ alias });

  if (!user) {
    throw new ModelError('Not found', 'user', alias);
  }
  return user;
};

export const getOrCreateUser = async ({ alias }) => {
  let user;
  try {
    user = await dbGetUserByAlias(alias);
  } catch (e) {
    if (e.message === 'Not found') {
      user = await addUser({ alias });
    }
  }
  return user;
};

export const delegatePoints = async ({ amount, sender, receiver }) => {
  // Verify sender is present and has enough points
  if (!sender) throw Error('No sender specified');
  // Throw error if sender is not present
  const senderUser = await dbGetUserByAlias(sender);

  const balance = await getUserBalance(sender);

  if (balance < amount)
    throw Error(
      `Naaah, you don't have so much points. Your current point balance is ${balance}`,
    );

  // Get receiver if not present create a new one
  const receiverUser = await getOrCreateUser({ alias: receiver });

  // Create points and pass to user
  const points = new Point({
    amount,
    used: false,
    delegatee: senderUser,
    owner: receiverUser,
  });

  await points.save();

  return points;
};

export const getUserBalance = async alias => {
  const colony = await getColonyById(0);
  const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];
  const { balance } = await getCurrentUserPeriod(alias, currentPeriodId);
  return balance;
};

export const setUserAlias = async ({ user, alias, blockNumber }) => {
  user.alias = alias;
  user.aliasSet = blockNumber;

  user.save(err => {
    if (err) return null;
  });

  return user;
};

export const addDomainToUser = async (req, res) => {
  // TODO:
};

export const addTaskToUser = async (req, res) => {
  // TODO:
};

export const changeUserAlias = async (req, res) => {
  // TODO:
  // this.getUserByAlias(req.body.alias);
};

export const updateUserAddress = async ({ alias, address }) => {
  if (!web3.utils.isAddress(address))
    throw Error('Address is incorrect, try again');
  const user = await dbGetUserByAlias(alias);
  if (!user) throw Error('Get some tips to get started');
  if (user.address)
    throw Error('Address is already set, I cannot change it by now');

  user.address = address;
  await user.save();
  return user;
};
