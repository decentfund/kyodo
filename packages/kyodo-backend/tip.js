import map from 'lodash/map';
import sum from 'lodash/sum';
import { Tip, Domain, getDomainByCode, Colony, getColonyById } from './db.js';
import {
  changeUserBalance,
  currentPeriod,
  getUserByAddressInPeriod,
} from './period';
import { getUserBalance, dbGetUserByAlias, getOrCreateUser } from './user';
import { dbCreateTask as createTask, getTaskByTitle } from './task';

export const sendTip = async (req, res) => {
  // TODO: colonyClient integration
  // transfer.send({ destinationAddress, amount }, options)
  // transferFrom.send({ sourceAddress, destinationAddress, amount }, options)

  let tipAmount = req.body.amount;
  if (tipAmount <= 0) {
    res.status(400).send('Seriously?!');
    return 'No money no honey, sorry';
  }

  let sender = await getUserByAddressInPeriod(req.body.from);
  let receiver = await getUserByAddressInPeriod(req.body.to);
  let senderAddress = sender[0].address;
  let senderBalance = sender[0].balance;
  let receiverAddress = receiver[0].address;

  if (!checkBalance(senderBalance, tipAmount)) {
    res.status(400).send('Inssuficient funds for the tip');
    return 'No money no honey, sorry';
  }

  await changeUserBalance(senderAddress, tipAmount);
  await changeUserBalance(receiverAddress, -tipAmount);

  let tip = new Tip({
    from: senderAddress,
    to: receiverAddress,
    amount: tipAmount,
    taskId: req.body.taskId,
    domainId: req.body.domainId,
    potId: req.body.potId,
    dateCreated: Date.now(),
    periodId: currentPeriod,
  });

  tip.save(err => {
    if (err) return console.error(err);
  });
  res
    .status(200)
    .send(
      `User ${senderAddress} Successfully tipped user ${receiverAddress} with ${tipAmount} `,
    );
};

export const sendNewTip = async ({
  sender,
  amount = 0,
  domain: code,
  title,
  receiver,
} = {}) => {
  try {
    const colony = await Colony.findOne();
    const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];

    // Verify sender is specified
    if (!sender) throw Error('No sender specified');

    // Throw error if amount is not specified
    if (amount <= 0) throw new Error('No money no honey, sorry');

    // Throw error if sender is not present
    const senderUser = await dbGetUserByAlias(sender);

    // Get receiver if not present create a new one
    const receiverUser = await getOrCreateUser({ alias: receiver });

    const balance = await getUserBalance(sender);
    if (!checkBalance(balance, amount))
      throw Error(
        "You have 0 points. But don't worry! You can earn them by making contributions other people find useful. Or just by making me laugh.",
      );

    // Getting domain
    // TODO: Throw error if domain is not found
    const domain = await getDomainByCode(code);

    // TODO: Try to find task
    // FIXME: Find task in period
    // If not available create a new one
    let task = await getTaskByTitle(title);
    if (!task) {
      task = await createTask({ title });
    }

    let tip = new Tip({
      from: senderUser,
      to: receiverUser,
      amount,
      task,
      domain,
      // potId: req.body.potId,
      dateCreated: Date.now(),
      // FIXME: Get period from colony
      periodId: currentPeriodId,
    });

    await tip.save();
    return tip;
  } catch (e) {
    if (e.name === 'SystemError') {
      throw new Error(
        'Uh-oh! Something went wrong! Please contact admin from the chat and submit a ticket with details in the repo http://github.com/decentfund/kyodo',
      );
    }
    if (e.message === 'Not found') {
      if (sender === e.value) {
        throw Error('Sender is not registered');
      }
    }
    throw e;
  }
};

export const getColonyCurrentPeriodId = async () => {
  const colony = await getColonyById(0);
  const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];
  return currentPeriodId;
};

export const getAllTips = async () => {
  const currentPeriodId = await getColonyCurrentPeriodId();
  let tips = await Tip.find({ periodId: currentPeriodId })
    .populate('task', 'taskTitle')
    .populate('domain', 'domainTitle')
    .populate('from')
    .populate('to');
  return tips;
};

export const getUserTips = async ({ user, periodId, direction }) => {
  const tips = await Tip.find(
    { periodId: periodId, [direction]: user._id },
    err => {
      if (err) return console.error(err);
    },
  )
    .populate('task', 'taskTitle')
    .populate('domain', 'domainTitle')
    .populate('from')
    .populate('to');

  return tips;
};

export const getAllUserTips = async ({ user, periodId = null }) => {
  const _periodId = periodId || (await getColonyCurrentPeriodId());
  const fromTips = await getUserTips({
    user,
    periodId: _periodId,
    direction: 'from',
  });
  const toTips = await getUserTips({
    user,
    periodId: _periodId,
    direction: 'to',
  });
  return [...fromTips, ...toTips];
};

export const getAllTipsInDomain = async domain => {
  const govDomain = await Domain.find({ domainTitle: domain });
  const tips = await Tip.find({ domain: govDomain });
  return {
    total: sum(map(tips, 'amount')),
    tips,
  };
};

export const checkBalance = (balance, amount) => {
  if (amount <= balance && amount > 0) return true;
  return false;
};
