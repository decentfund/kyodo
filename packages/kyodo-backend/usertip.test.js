import 'babel-polyfill';
import { sendNewTip } from './tip';
import pick from 'lodash/pick';
import map from 'lodash/map';
import { getAllUserTips, getAllTipsInDomain } from './tip';
import { Domain, Colony } from './db';
import { addUserWithBalance } from './test/helpers/user';
import {
  connectMongoose,
  clearDb,
  disconnectMongoose,
  createColony,
  createDomain,
} from './test/helpers';

beforeAll(connectMongoose);
beforeEach(clearDb);
afterAll(disconnectMongoose);

describe('get user tips', () => {
  let insUser, insUser2;

  beforeEach(async () => {
    await createColony();
    await createDomain();

    insUser = await addUserWithBalance('igor');
    insUser2 = await addUserWithBalance('alina');

    // const users = await User.find();

    await sendNewTip({
      sender: insUser.alias,
      receiver: 'someone',
      amount: 2,
      title: 'yes',
      domain: 'GOV',
    });

    await sendNewTip({
      sender: insUser2.alias,
      receiver: insUser.alias,
      amount: 3,
      title: 'hey!',
      domain: 'GOV',
    });
  });

  it('sets test properly', async () => {
    const domains = await Domain.find();
    expect(domains.length).toBe(1);

    const colonies = await Colony.find();
    expect(colonies.length).toBe(1);
  });

  it('returns all user tips', async () => {
    const userTips = await getAllUserTips({ user: insUser });
    expect(
      map(userTips, item =>
        pick(item, [
          'amount',
          'domain.domainTitle',
          'from.alias',
          'periodId',
          'task.taskTitle',
          'to.alias',
        ]),
      ),
    ).toMatchSnapshot();
  });

  it('returns all tips in domain', async () => {
    const { total, tips } = await getAllTipsInDomain('GOV');
    expect(total).toEqual(5);
    expect(
      map(tips, item => pick(item, ['amount', 'periodId'])),
    ).toMatchSnapshot();
  });
});
