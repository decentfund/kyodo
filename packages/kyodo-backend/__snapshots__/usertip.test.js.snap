// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get user tips returns all tips in domain 1`] = `
Array [
  Object {
    "amount": 2,
    "periodId": 0,
  },
  Object {
    "amount": 3,
    "periodId": 0,
  },
]
`;

exports[`get user tips returns all user tips 1`] = `
Array [
  Object {
    "amount": 2,
    "domain": Object {
      "domainTitle": "GOV",
    },
    "from": Object {
      "alias": "igor",
    },
    "periodId": 0,
    "task": Object {
      "taskTitle": "yes",
    },
    "to": Object {
      "alias": "someone",
    },
  },
  Object {
    "amount": 3,
    "domain": Object {
      "domainTitle": "GOV",
    },
    "from": Object {
      "alias": "alina",
    },
    "periodId": 0,
    "task": Object {
      "taskTitle": "hey!",
    },
    "to": Object {
      "alias": "igor",
    },
  },
]
`;
